File: data/derived_cot/rq3_prompting/results/translation/glm-5.1/cpp_to_py/code_output/Warehouse.py
class Warehouse:
    def __init__(self):
        self._inventory = {}
        self._orders = {}

    def add_product(self, product_id, name, quantity):
        if product_id not in self._inventory:
            self._inventory[product_id] = {
                "name": name,
                "quantity": str(quantity)
            }
        else:
            current_quantity = int(self._inventory[product_id]["quantity"])
            self._inventory[product_id]["quantity"] = str(current_quantity + quantity)

    def update_product_quantity(self, product_id, quantity):
        if product_id in self._inventory:
            current_quantity = int(self._inventory[product_id]["quantity"])
            self._inventory[product_id]["quantity"] = str(current_quantity + quantity)

    def get_product_quantity(self, product_id):
        if product_id in self._inventory:
            return int(self._inventory[product_id]["quantity"])
        return 0

    def create_order(self, order_id, product_id, quantity):
        available_quantity = self.get_product_quantity(product_id)
        if available_quantity >= quantity:
            self.update_product_quantity(product_id, -quantity)
            self._orders[order_id] = {
                "product_id": str(product_id),
                "quantity": str(quantity),
                "status": "Shipped"
            }
            return True
        return False

    def change_order_status(self, order_id, status):
        if order_id in self._orders:
            self._orders[order_id]["status"] = status
            return True
        return False

    def track_order(self, order_id):
        if order_id in self._orders:
            return self._orders[order_id]["status"]
        return ""

    def orders(self):
        return self._orders

    def inventory(self):
        return self._inventory

File: data/derived_cot/rq3_prompting/results/translation/qwen/py_to_cpp/code_output/Warehouse.cpp
#include <iostream>
#include <unordered_map>
#include <variant>
#include <string>

// Define structs for Product and Order
struct Product {
    std::string name;
    int quantity;
};

struct Order {
    int product_id;
    int quantity;
    std::string status;
};

class Warehouse {
private:
    std::unordered_map<int, Product> inventory;
    std::unordered_map<int, Order> orders;

    bool is_product_exists(int product_id) const {
        return inventory.find(product_id) != inventory.end();
    }

    int get_product_quantity(int product_id) const {
        return inventory.at(product_id).quantity;
    }

public:
    Warehouse() = default;

    void add_product(int product_id, const std::string& name, int quantity) {
        if (!is_product_exists(product_id)) {
            inventory[product_id] = {name, quantity};
        } else {
            inventory[product_id].quantity += quantity;
        }
    }

    void update_product_quantity(int product_id, int quantity) {
        if (is_product_exists(product_id)) {
            inventory[product_id].quantity += quantity;
        }
    }

    std::variant<int, bool> get_product_quantity(int product_id) {
        if (is_product_exists(product_id)) {
            return inventory[product_id].quantity;
        }
        return false;
    }

    bool create_order(int order_id, int product_id, int quantity) {
        if (get_product_quantity(product_id) >= quantity) {
            update_product_quantity(product_id, -quantity);
            orders[order_id] = {product_id, quantity, "Shipped"};
            return true;
        }
        return false;
    }

    bool change_order_status(int order_id, const std::string& status) {
        if (orders.find(order_id) != orders.end()) {
            orders[order_id].status = status;
            return true;
        }
        return false;
    }

    std::variant<std::string, bool> track_order(int order_id) {
        if (orders.find(order_id) != orders.end()) {
            return orders[order_id].status;
        }
        return false;
    }
};